### ISSUE TITLE
[Bug] spoon reorders record parameters

### ISSUE THREAD

**Issue by xzel23** [INRIA/spoon#4365]
**Describe the bug**
Record parameters are output in the wrong order, i. e. first and second are switched in this example (first line of code).

**To Reproduce**

Input:
```java
record Pair(int first, float second) {
    public static Pair of(int first, float second) {
        return new Pair(first, second);
    }
}
```

Processing with Spoon:
```java
java -classpath ${DIR}/lib/*.jar spoon.Launcher -i "${DIR}/issue" -o "${DIR}/issue/processed" --compliance 16
```

Output:
```java
record Pair(float second, int first) {
    public static Pair of(int first, float second) {
        return new Pair(first, second);
    }
}
```

Note that you can also open a pull request reproducing the issue and reference that, instead of writing additional information here.

**Operating system, JDK and Spoon version used**

* OS: MacOS 12.1
* JDK: GraalVM CE 21.3 (JDK 17.0.1)
* Spoon version: spoon-core-10.0.1-beta-2-jar-with-dependencies.jar

UPDATE: I changed the second parameter so that it can be observed that not just the names but also the types are swapped.

**Comment by MartinWitt** [comment]
It seems like using a HashSet for insertion ordered elements is still bad. Because `Hashset` related bugs are pretty hard to reproduce locally without exactly the same JVM and OS setup, could you try if changing the https://github.com/INRIA/spoon/blob/c65b2eae4984949ca4ecc2641cc7f4052149935e/src/main/java/spoon/support/reflect/declaration/CtRecordImpl.java#L45 `HashSet` here to a `LinkedHashSet` fixes the issue?

**Comment by xzel23** [comment]
What? Of course that's bad. Order of elements in a set is undefined. If you need order, you need a LinkedHashSet. In that case, I would also recommend changing the declaration to `LinkedHashSet` as the order of elements is not an implementation detail but part of the contract. If you want, I can create a test case, apply the fix and create a PR (if it works).

**Comment by MartinWitt** [comment]
The problem is, we try to be as backwards compatible as possible, and therefore we can't change the method parameters any more. The class itself only calls add and remove operations on the set, so changing the dynamic type to `LinkedHashSet` should be the best solution.  
> If you want, I can create a test case, apply the fix and create a PR (if it works).

Sure, that would be awesome.

**Comment by xzel23** [comment]
The good news is: there's already a test case for this, `testMultipleParameterRecord()`, it just didn't test for the order of components. Fixed that, and voilà: test fails with the order of elements in the declaration being wrong.

Now the bad news: changing components to a LinkedHashSet doesn't help. I'll dig into this.

**Comment by xzel23** [comment]
TLDR; CloneHelper cannot clone CtRecord correctly because the cloning functionality has been moved to a place where implementation details are not known. Can be fixed by always cloning sets to LinkedHashSet, even where that is not optimal.

---

Sorry for the rant, but here are the details:

OK, using LinkedHashSet fixes part of the problem. Everything works as long as long as the CtRecord stays the same. But for whatever reason, spoon is cloning elements all over the place. The problems with Java's clone() implementation have been know for ages, so in spoon, it obviously got replaced by something that is even worse: a CloneHelper class that probably was introduced around the time the GOF book was famous and everything had to be done using the visitor pattern, even where it is not really necessary. So we end up here:

```
	public <T extends CtElement> Set<T> clone(Set<T> elements) {
		if (elements instanceof EmptyClearableSet) {
			return elements;
		}
		if (elements == null || elements.isEmpty()) {
			return EmptyClearableSet.instance();
		}
		Set<T> others = new HashSet<>(elements.size());
		for (T element : elements) {
			addClone(others, element);
		}
		return others;
	}

```

The problem is that the components member is cloned, which should be easy: use instanceOf to check for LinkedHashSet and construct the correct type of Set.

But CloneHelper is passed the original objects components wrapped in a Collections.unmodifiableSet(), which is a good thing. But a LinkedHashSet that is wrapped by unmodifiableSet() is indistinguishable (apart from using reflection and other tricks) from other implementations of the Set interface.

CloneHelper cannot do the right thing here! A simple solution would be to just always return a LinkedHashSet, but I won't even try that because it's a really bad hack.

As you mentioned backwards compatibility: I think that every once in a while, breaking changes have to be made to keep a growing codebase in a maintainable state, and getting rid of CloneHelper (and possibly the need to clone/copy elements in so many places of the code) would be a change that makes spoon more maintainer-friendly.

Just think about what happens here: you want to clone a ctRecordImp, and this is what happens (just for the components member):
 - an UnmodifiableSet is created as a wrapper around the components member of the original instance.
 - CloneHelper is used to clone the UnmodifableSet and all of its members
 - the result is passed to CtRecordImpl.setRecordComponents()
 - CtRecordImpl clears its own component member by calling removeRecordComponent for every contained element
 - for each element in the cloned set, addRecordComponent is called

This is IMHO unneeded overhead. Also things that should be implementation details of CtRecordImpl have been moved to CloneHelper which now depends on each and every implementation of the classes it can clone, and those classes in turn depend on CloneHelper.

OK, so this works:
```java
	@java.lang.Override
	public void visitCtRecord(spoon.reflect.declaration.CtRecord recordType) {
		spoon.reflect.declaration.CtRecord aCtRecord = recordType.getFactory().Core().createRecord();
		this.builder.copy(recordType, aCtRecord);
		aCtRecord.setAnnotations(this.cloneHelper.clone(recordType.getAnnotations()));
		aCtRecord.setSuperInterfaces(this.cloneHelper.clone(recordType.getSuperInterfaces()));
		aCtRecord.setTypeMembers(this.cloneHelper.clone(recordType.getTypeMembers()));
		aCtRecord.setFormalCtTypeParameters(this.cloneHelper.clone(recordType.getFormalCtTypeParameters()));
		aCtRecord.setRecordComponents(new LinkedHashSet<>(recordType.getRecordComponents().stream().map(e -> cloneHelper.clone(e)).collect(Collectors.toList())));
		aCtRecord.setComments(this.cloneHelper.clone(recordType.getComments()));
		this.cloneHelper.tailor(recordType, aCtRecord);
		this.other = aCtRecord;
	}
```

But only until `CtGenerationTest.testGenerateCloneVisitor()` detects that I have changed above method. So I am back to always returning a LinkedHashSet. PR created. If anybody has a better solution, please let me know.

**Comment by SirYwell** [comment]
Not sure how we failed to see that earlier, but a `Set` really isn't a good choice here.

I'd say it would be worth to deprecate (but keep) the methods with `Set` in `CtRecord`. For the getter, we'd need something like `getRecordComponentList()`, otherwise we'd have clashing signatures. For the setter, overloading works fine, so that shouldn't be an issue.

The old getter can be annotated with `@DerivedProperty` and the implementation can just use `Set.copyOf(...)`. The setter can be implemented the same way the current method is implemented.

WDYT?

**Comment by xzel23** [comment]
Sounds good to me. I will change it and see if I run into other problems.

**Comment by xzel23** [comment]
OK, so it is not as easy as that. Because getRecordComponents() returns a Set, everywhere where that is used, the order of elements is undefined again. I tried it out, and record parameters were garbled again in my test case. So I'd rather leave it like this and suggest that the components member is changed to List<> in the next major version. I also suggest changing setRecordComponents to just accept Collection<> and not List<> or Set<>, as that will work with both.

**Comment by xzel23** [comment]
Fixed after applying https://github.com/INRIA/spoon/pull/4366.
